MarcusBarnes/mik
# Issue 240

**Github issue**: https://github.com/MarcusBarnes/mik/issues/240

# What does this Pull Request do?

Adds input validators, as described in #240, for CSV Single File, CSV Newspapers, Csv Compound, and Csv Books toolchains

# What's new?

* added working input validators for CSV Single File, CSV Newspapers, Csv Books, and Csv Compound toolchains
* modified mik to integrate the validators in 'strict' mode
* modified the writers for those three toolchains to integrate the validators in 'realtime' mode
* added PHPUnit tests and test assets for all four input validators
* modified .travis.yml to run the tests on these input validators separately from the other PHPUnit tests

# How should this be tested?

## PHPUnit tests

Running `phpunit --group inputvalidators --bootstrap vendor/autoload.php tests` in the MIK directory will run tests for all four input validators.

## Manual tests

* For all four toolchains, adding `validate_input = false` to the .ini file's `[FILE_GETTER]` section will disable the input validators. MIK's behavior will be the the same as it was before the addition of the code in this PR.
* For all four toolchains, adding `validate_input_type = strict` to the .ini file's `[FILE_GETTER]` section will cause MIK to validate the input packages and not generate any output. In other words, with this configuration setting, running `./mik -c foo.ini` will not produce any Islandora ingest packages. All it will produce is a log (as defined in detailing any validation errors. The validation log's path is defined in the optional ['LOGGING']['path_to_input_validator_log'] .in setting; default location is input_validator.log in the MIK output directory.
* For all toolchains, adding `validate_input_type = realtime` (the default; will be in effect if no configuration option is provided) to the .ini file's `[FILE_GETTER]' section will cause MIK to validate the input packages as they are being generated. If a package fails any input validation test, its generation will be skipped and its ID will be added to the problem.log. All validation failures will be detailed in the input validator log.
* Toolchain-specific tests:
  * CSV single file
    * The only check we make is that the file named in the CSV record's file_name_field exists
  * CSV compound
    * Compound object-level directory named in the CSV record's compound_directory_field must exist, and be a directory
    * Compund object-level directory cannot contain 'Thumbs.db', '.Thumbs.db', 'DS_Store', or '.DS_Store'
    * Compound object-level directory must contain at least n files, where n is defined in $settings['WRITER']['min_children']; (as per #291). Default is 2 files.
    * Files in compound object-level directory must be named such that their last filename segment is numeric.
  * CSV newspapers
    * Issue-level directory named in the CSV record's file_name_field must exist and be named using the yyyy-mm-dd pattern.
    * Issue-level directory must contain at least one page file.
    * The page filenames must be names such that their page sequence numbers are numeric.
  * CSV books
    * Book-level directory named in the CSV record's file_name_field must exist, and be a directory.
    * The book-level directory cannot be empty.
    * The book-level directory cannot contain directory cannot contain 'Thumbs.db', '.Thumbs.db', 'DS_Store', or '.DS_Store'.
    * Files in the book-level directory must have one of the allowed extensions ('tiff', 'tif', or 'jp2'.).
    * Files in the book-level directory must be named such that their page sequence numbers are numeric.

Given that there are PHPUnit tests, I think it's OK to do a manual smoke test to see if the configuration options work the way you expect, using the CSV single file toolchain.

There is one issue that I cannot figure out, but that is harmless: in 'realtime' mode, the "Initializing log" entry shows up twice in input_validator.log. I can open a separate issue for this later.
